### ISSUE TITLE
A simple mechanism for bandwidth accounting

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#5623]
#4719 discussed the excellent insight that bandwidth accounting does not require hard security requirements for our upcoming economy. In particular, an adequate reputation mechanism already (partially) addresses many security concerns, including the Sybil Attack and fake transaction creation.

To this end, @grimadas and I have devised a very simple, yet effective approach to account bandwidth contributions in Tribler. The upcoming algorithm does not impose a personal ledger structure, therefore significantly simplifying the design and implementation. A compelling advantage is the storage overhead of our algorithm is kept to a minimum. At the minimum, individuals are only required to store the last transaction for every counterparty they have interacted with. Each transaction contains two signatures, one from each counterparty. Each transaction represents a single edge in the work graph and contains the weight of the edge between the interacting parties. Transactions contain an ever-growing counter, indicating the total amount of bandwidth token with a specific counterparty. Therefore, transactions can easily be merged when a state inconsistency is detected. That's all there is.

A key design decision is that transactions will be relatively small, e.g., at most 50MB. Therefore, it is not a big problem if a counterparty goes offline, or when a peer refuses to sign a specific transaction. Like TrustChain, liveness is guaranteed in the presence of network churn, and users can engage in multiple transactions at the same time.

Implementing this approach allows us to get rid of the TrustChain implementation. TrustChain will move to the AnyDex repository since it is still used to account trades in the decentralized market.

## The approach

The following algorithm is an essential step towards a secure micro-economy around bandwidth transfers. It requires minimal computation and state storage overhead. We exemplify our accounting approach using a simple example, where party A payouts 20MB to party B. It consists of the following seven steps:

```
1) A creates  T = (sig_a, EMPTY_SIG, total_taken_a_from_b = db.total_taken_a_from_b + 20)
2) A sends T to B
3) B verifies T
	3a) verify truthfulness of data in T
	3b) verify validity of signature sig_a
	3c) compare T against the last known state of A => total_taken_a_from_b > db.total_taken_a_from_b must hold
4)
	4a) If T is valid, create T’ = (sig_a, sig_b, total_taken_a_from_b) and commit to DB
	4b) If T is not valid, send the last (dual-signed) T’ to A
5) send T’ it to A
6) A verifies T’
	6a) verify validity of signature sig_a and sig_b
	6b) check that T’ is the last state => total_taken_a_from_b > db.total_taken_a_from_b
7)
	7a) if T’ is valid, commit to DB
	7b) if T’ is not valid, ignore T’
```

## Reconstructing the work graph

To devise input for the reputation mechanism, we require peers to reconstruct the work graph. This can easily be achieved by requesting the set of latest transactions  from a counterparty. This collection process can run at a very low speed (with minimal impact on other components in Tribler), and with low computational overhead. Eventually, each peer will build the work graph and devise a ranking of peers in the network through local computations.

## Subjective reputation mechanism

In parallel to the implementation of the above algorithm, an efficient reputation mechanism will be implemented. This progress will be tracked in a separate issue.

This ticket will be updated with results/issues arising as the implementation progresses.

**Comment by devos50** [comment]
A first implementation can be found in https://github.com/Tribler/tribler/pull/5626. I aimed to keep this implementation as simple as possible for the moment. This PR needs a bit more polish to be considered ready for review. The overall structure of the mechanism resembles TrustChain, but without many unneeded features and functionalities (e.g., a linear chain structure).

While implementing the algorithm as discussed in this issue was straightforward, I ran into a few issues regarding integration in Tribler. Specifically:
- The trust statistics graph is using TrustChain blocks to visualise the work graph and bandwidth transfers between peers. This likely requires some refactoring as the TrustChain backend will be unused after merging this PR. @xoriole since you are the original author of this graph, do you want to help out with this? Note that this can be done after merging this PR (The trust screen should still work for the moment being since I did not remove anything TrustChain-related).
- The tunnels are using a different payout method. This leads to incompatibility issues when performing payouts between Tribler instances. This might require a community upgrade of the tunnels. @egbertbouman what is your opinion on this?

**Comment by synctext** [comment]
> individuals are only required to store the last transaction for every counterparty they have interacted with.

This is fascinating. I can see two design directions. What do we keep in a transaction with a counterparty? It can be the latest transaction, over-writing prior interactions we had. Another approach is that each peer keeps long-term counters with their counterparties. You drop the old transaction, update the counter, and publish only the latest block.

**Comment by devos50** [comment]
> Another approach is that each peer keeps long-term counters with their counterparties. You drop the old transaction, update the counter, and publish only the latest block.

I think this is exactly what we are doing right now. Each peer only stores the last transaction in which it gave bandwidth, for each counterparty. Note that the counter in a transaction with a counterparty is an ever-growing number. To resolve conflicts and out-of-order delivery of messages, a counterparty only signs an incoming sign request when the counter in the received transaction is strictly greater than the locally known one (in the database). This approach keeps the storage overhead and computational complexity to a minimum.

> How are we going to do tunnel payouts?

Currently, we do this the same way as TrustChain did it - each hop payouts the next hop in the circuit. That said, there is still much room for improvement here, but that is left for a future iteration.

**Comment by synctext** [comment]
boostrap idea: download 25MByte of work graph with https://en.wikipedia.org/wiki/Giant_component

**Comment by xoriole** [comment]
>The trust statistics graph is using TrustChain blocks to visualise the work graph and bandwidth transfers between peers. This likely requires some refactoring as the TrustChain backend will be unused after merging this PR. @xoriole since you are the original author of this graph, do you want to help out with this? Note that this can be done after merging this PR (The trust screen should still work for the moment being since I did not remove anything TrustChain-related).

Sure, I'd love to work on this.

**Comment by devos50** [comment]
#5626 has been merged 👍 . I will create a new issue for the trust graph that should be fixed.

Also, I will update the description of the token economy to include our basic algorithm.
